File: days/01/ts/main.ts
import path from "node:path";
import fs from "node:fs/promises";

class BST {
  root: Node | undefined;

  constructor() {}
  insert(value: number, meta?: Record<string, unknown>) {
    let current: Node | undefined;
    let x = this.root;
    const next = new Node(value, meta);

    while (x) {
      current = x;
      if (next.data < x.data) {
        x = x.left;
      } else {
        x = x.right;
      }
    }

    next.parent = current;
    if (!current) {
      this.root = next;
    } else if (next.data < current.data) {
      current.left = next;
    } else {
      current.right = next;
    }
  }
}
class Node {
  data: number;
  meta?: Record<string, unknown>;
  parent?: Node;
  left?: Node;
  right?: Node;
  constructor(value: number, meta?: Record<string, unknown>) {
    this.data = value;
    if (meta) {
      this.meta = meta;
    }
  }
}

async function main() {
  const datapath = path.resolve(__dirname, "../input.txt");
  const file = await fs.open(datapath);

  let elves = new BST();
  let index: number = 0;
  let calories = 0;

  for await (const chunk of file.readLines()) {
    // empty line implies new elf
    if (!chunk) {
      elves.insert(calories, { index });
      calories = 0;
      index += 1;
      continue;
    }

    let value = parseInt(chunk);
    calories += value;
  }

  let total = 0;
  let t = 3;
  function walk(node?: Node) {
    if (node) {
      walk(node.right);
      if (t > 0) {
        total += node.data;
      }
      t = t - 1;
      walk(node.left);
    }
  }
  walk(elves.root);
  return total;
}

main().then((total) => {
  console.log(`top three elves total calories: ${total}`);
});

File: days/01/rust/src/main.rs
use std::fs::File;
use std::io::{self, prelude::*, BufReader};

fn main() -> io::Result<()> {
    let file = File::open("input.txt")?;
    let reader = BufReader::new(file);

    let mut elves = Vec::new();
    let mut calories: i32 = 0;

    for line in reader.lines() {
        let _: i32 = match line?.trim().parse() {
            Ok(num) => {
                calories += num;
                num
            },
            Err(_) => {
                elves.push(calories);
                calories = 0;
                continue
            },
        };
    }

    elves.sort_by(|a, b| b.cmp(a));

    let mut total: i32 = 0;
    for i in 0..3 {
        total += elves[i];
    }

    println!("top three total: {}", total);
    println!("largest: {}", elves[0]);

    Ok(())
}